src/box/iproto.cc

```diff
@@ -1857,7 +2069,46 @@ net_send_msg(struct cmsg *m)
 {
 	struct iproto_msg *msg = (struct iproto_msg *) m;
 	struct iproto_connection *con = msg->connection;
+	struct iproto_stream *stream = msg->stream;
+	struct iproto_msg *tmp;
+
+	if (stream == NULL)
+		goto send_msg;
+
+	tmp = stailq_shift_entry(&stream->pending_requests,
+				 struct iproto_msg, in_stream);
+	assert(tmp == msg);
+	(void)tmp;
+	errinj_stream_msg_count_add(-1);
+
+	if (stailq_empty(&stream->pending_requests)) {
+		struct mh_i64ptr_node_t node = { stream->id, NULL };
+		mh_i64ptr_remove(con->streams, &node, 0);
+		iproto_stream_delete(stream);
+	} else {
+		struct cmsg *m;
+		if (!evio_has_fd(&con->input)) {
+			m = &stream->on_disconnect;
+			struct iproto_thread *td = con->iproto_thread;
+			cmsg_init(m, td->destroy_stream_on_disconnect_route);
```

Comment:
AFACS you discard all queued requests when the connection is closed by the client, but I think we should process them. Am I missing something?

Anyway, it'd be nice to see a comment here.

Reply:
Why not do it in this patch? AFAICS currently iproto processes all requests even if the peer connection is closed (please amend me if I'm wrong). Streamed messages behave differently - they are discarded. This is confusing. It shouldn't be too difficult to process them all before shutdown.

Reply:
The problem of trying to finish all the requests is that it does not give you any guarantees until proper graceful shutdown is implemented. See #5924. You need a protection against the server closing the socket too early before the client receives all the responses. Not only the server needs to finish the requests, but also the client must confirm that all the responses are delivered. Otherwise the client can't tell which of its requests are finished and whose result is unknown. And that makes the attempt to finish all the requests pointless if the connection is already being closed.

For this patchset we agreed to do what is easier to implement. It appeared that it is easier to discard everything what was not started yet. We didn't sacrifice any guarantees by that. If there is a way to simplify the code by finishing all the requests, I wouldn't mind.

Reply:
I don't think it's related to graceful shutdown. I think a user may want to send some requests and immediately close the connection without waiting for a result. This doesn't give the user any guarantees, but if this is about logging some events or stats, no guarantees are required - they just want the requests to succeed unless there's a server failure. Without streams, this will work just fine. However, if the user attempts to use streams, requests might get silently dropped upon closing connection, even if the server is perfectly healthy.

Reply:
I have reworked this part of the patch. But, as I wrote in a private discussion, we cannot guarantee the processing of all sent requests, when closing the connection, if net_msg_max is reached during processing requests. Therefore, for streams, I just keep the guarantees that are currently available for messages not from the stream.